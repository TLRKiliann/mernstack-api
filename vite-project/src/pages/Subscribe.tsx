import React, { useState } from 'react'
import serviceRouting from '../services/serviceRouting'
import usePersonnalHook from '../hook/personnal.hook'
import { useAuthLogin } from '../context/AuthProvider'
import { UserType } from '../models/usertype'
import { useNavigate } from 'react-router-dom'
import '../stylePages/Subscribe.scss'

type Field = {
  value?: any
  label?: string
  error?: string
  isValid?: boolean
}

type Form = {
  firstName: Field
  lastName: Field
  password: Field
  age: Field
  email: Field
  location: Field
  gender: Field
}

const options: Field[] = [
  {
    label: "Male",
    value: "Male",
  },
  {
    label: "Female",
    value: "Female",
  }
]

const Subscribe: React.FC = () => {

  const Navigate = useNavigate()

  const [form, setForm] = useState<Form>({
    firstName: {value: ''},
    lastName: {value: ''},
    password: {value: ''},
    age: {value: 20},
    email: {value: ''},
    location: {value: ''},
    gender: {value: 'Male'}
  })

  const users = usePersonnalHook()
  //console.log(datas, "datas")
  const [message, setMessage] = useState<string>("")
  const [datas, setDatas] = useState<Array<UserType>>([])

  const handleInputChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
    const fieldName: string = e.target.name;
    const fieldValue: string | number | boolean = e.target.value;
    const newField: Field = {[fieldName]: {value: fieldValue}}

    setForm({...form, ...newField});
  }

  const validateFormSub = () => {
    let newForm: Form = form;

    //Validate firstname
    if (form.firstName.value.length < 3) {
      const errorMsg: string = "Votre prÃ©nom doit faire minimum 3 charactÃ¨res";
      const newField: Field = {value: form.firstName.value, error: errorMsg, isValid: false};
      newForm = {...newForm, ...{ firstName: newField }};
    } else {
      const newField: Field = {value: form.firstName.value, error: '', isValid: true};
      newForm = {...newForm, ...{ firstName: newField }};
    }

    //Validate password
    if (form.password.value.length < 6) {
      const errorMsg: string = "Votre mot de passe doit faire au moins 6 charatÃ¨res"
      const newField: Field = {value: form.password.value, error: errorMsg, isValid: false};
      newForm = {...newForm, ...{ password: newField }};
    } else {
      const newField: Field = {value: form.password.value, error: '', isValid: true};
      newForm = {...newForm, ...{password: newField}}
    }

    setForm(newForm);
    return newForm.firstName.isValid && newForm.password.isValid;
  }

  const generateId = () => {
    const maxId: number = users.length > 0
      ? Math.max(...users?.map(d => d.id))
      : 0
    return maxId + 1;
  };

  const handleValidateSub = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault()
    const isFormValid = validateFormSub();
    if (isFormValid) {
      setMessage("ðŸ‘‰  Registration en cours ...")

      const newMember = {
        id: generateId(),
        img: "http://localhost:5173/src/assets/snapface/ronaldo.png",
        firstName: form.firstName.value,
        lastName: form.lastName.value,
        age: form.age.value,
        email: form.email.value,
        location: form.location.value,
        gender: form.gender.value,
        mainroom: "",
        room: "",
        isConnected: false,
        signalRecieve: false,
        sentMsg: "",
        messagebox: "",
        returnConfirm: false
      }
      
      console.log(newMember, "newMember")

      serviceRouting
        .createMember(newMember)
        .then(initialData => {
          setDatas(users.concat(initialData))
        })
        .catch((err) => {
          console.log("Error during creation of new Member !")
          setDatas([])
        })
      Navigate('/login')
    } else {
      setMessage("something went wrong")
    }
  }

  return (
    <div className="subscribe">

      <form
        className="form--subscribe"
        autoComplete="off"
        onSubmit={(e) => handleValidateSub(e)}
      >
        <div className="div--subscribe">
          <h1 className="title--framesub">Subscribe</h1>

          {message && <div className="message--sub">
            {message}
            </div>
          }

          <label className="lbl--sub" htmlFor="sub">
            Firstname
          </label>
          <input
            type="text"
            name="firstName"
            value={form.firstName.value}
            onChange={(e) => handleInputChange(e)}
            placeholder="firstname"
            className="input--sub"
            required
          />

          {form.firstName.error && 
            <div 
              className="firstname--error">{form.firstName.error}
            </div>
          }

          <label className="lbl--sub" htmlFor="sub">
            Lastname
          </label>
          <input
            type="text"
            name="lastName"
            value={form.lastName.value}
            onChange={(e) => handleInputChange(e)}
            placeholder="lastname"
            className="input--sub"
            required
          />

          <label className="lbl--sub" htmlFor="sub">
            Password
          </label>
          <input
            type="password"
            name="password"
            value={form.password.value}
            onChange={(e) => handleInputChange(e)}
            placeholder="password"
            className="input--sub"
            required
          />

          {form.password.error && 
            <div 
              className="password--error">{form.password.error}
            </div>
          }

          <label className="lbl--sub" htmlFor="sub">
            Age
          </label>
          <input
            type="number"
            name="age"
            value={form.age.value}
            onChange={(e) => handleInputChange(e)}
            min="20" max="120"
            className="input--sub"
            placeholder="age"
            required
          />

          <label className="lbl--sub" htmlFor="sub">
            Email
          </label>
          <input
            type="email"
            name="email"
            value={form.email.value}
            onChange={(e) => handleInputChange(e)}
            placeholder="email"
            className="input--sub"
            required
          />

          <label className="lbl--sub" htmlFor="sub">
            Location
          </label>
          <input
            type="text"
            name="location"
            value={form.location.value}
            onChange={(e) => handleInputChange(e)}
            placeholder="location"
            className="input--sub"
            required
          />

          <div className="div--subselect">
            <label htmlFor="gender" className="lbl--sub">
              Gender : 
            </label>
            <select
              id="gender"
              name="gender"
              value={form.gender.value}
              onChange={(e) => handleInputChange(e)}
              className="select--sub"
            >
            {options.map((option) => (
              <option
                key={option.value}
                name="gender"
                value={option.value}>
                  {option.label}
              </option>
            ))}
            </select>
          </div>

        </div>

        <button type="submit" className="btn--sub">
          Sign Up
        </button>

      </form>
    </div>
  )
}

export default Subscribe
