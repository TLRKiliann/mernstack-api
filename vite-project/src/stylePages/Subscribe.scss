.subscribe {
  position: absolute;
  top: 0px;
  right: 0px;
  left: 0px;
  min-height: 100vh;
  background: linear-gradient(180deg, purple, orange);
}

.title--subscribe {
  margin-top: 80px;
  margin-left: 20px;
  font-family: "Junction", sans-serif;
}

.form--subscribe {
  width: 480px;
  height: auto;
  padding: 20px;
  margin: auto;
  margin-top: 7%;
  margin-bottom: 12%;
  display: flex;
  flex-direction: column;
  border: 2mm ridge royalblue;
  background: rgba(45, 85, 255, 0.4);
  border-radius: 12px;

  .message--sub {
    margin: auto;
    margin-bottom: 20px;
    font-size: 1.1rem;
    color: lightgreen;
  }

  .firstname--error {

  }

  .password--error {
    
  }


  .div--subscribe {
    width: 300px;
    margin: auto;
    display: flex;
    flex-direction: column;
  }

  .title--framesub {
    margin: auto;
    margin-bottom: 20px;
    margin-left: 0px;
  }

  .lbl--sub {
    font-family: 'Junction';
    font-size: 1.2rem;
    font-weight: bold;
  }

  .input--sub {
    width: 300px;
    margin-top: 5px;
    margin-bottom: 10px;
    padding: 5px 5px;

    font-family: 'Junction';
    font-size: 1.0rem;
    font-weight: bold;

    border: 1mm inset rgba(45, 85, 255, 0.4);
    &::placeholder {
      color: lightgrey;
    }
  }

  .div--subselect {
    width: 100%;
    margin: auto;
    margin-top: -10px;
    margin-bottom: 10px;
    display: flex;
    flex-direction: row;
    align-items: center;

    .select--sub {
      width: 100px;
      margin: 15px 0px;
      margin-left: 10px;
      padding: 5px 10px;
    }
  }
  
  .btn--sub {
    margin: auto;
    padding: 10px 20px;
    
    font-family: 'Junction';
    font-size: 1.0rem;
    font-weight: bold;
    
    background: royalblue;
    border: 3px outset royalblue;
    border-radius: 12px;

    color: whitesmoke;
    &:hover {
      background-color: dodgerblue;
      border: 3px outset dodgerblue;
      cursor: pointer;
      transform: scale(1.1);
      color: lightblue;
    }
  }
}