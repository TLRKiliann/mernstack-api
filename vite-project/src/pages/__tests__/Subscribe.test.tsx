import React from 'react'
import {useNavigate} from 'react-router-dom'
import { test, expect, vi, beforeEach } from 'vitest'
import { fireEvent, render, screen } from '@testing-library/react'
import '@testing-library/jest-dom'
import Subscribe from "../Subscribe"
import handleInputChange from '../Subscribe'
import validateFormSub from '../Subscribe'
import handleValidateSub from '../Subscribe'
import generateId from '../Subscribe'

//useNavigate
useNavigate: () => ({
  navigate: vi.fn().mockImplementation(() => ({}))
})

vi.mock('react-router-dom', () => ({
  ...vi.importActual('react-router-dom'),
  useNavigate: () => mockedNavigate
}))

beforeEach(async (context) => {
  // extend context
  context.foo = 'bar'
})

test('should work', ({ foo }) => {
  console.log(foo) // 'bar'
})

test('Subscribe handleInputChange to be defined', () => {
  const funcInputChange = handleInputChange
  expect(funcInputChange).toBeDefined()
})

test('Subscribe validateFormSub to be defined', () => {
  const funcValid = validateFormSub
  expect(funcValid).toBeDefined()
})

test('Subscribe handleValidateSub to be defined', () => {
  const funcValidSub = handleValidateSub
  expect(funcValidSub).toBeDefined()
})

test('Subscribe generateId to be defined', () => {
  const funcGenId = generateId
  expect(funcGenId).toBeDefined()
})

test("Subscribe submit form test", () => {
  const onSubmit = vi.fn()
  const { getByTestId } = render(
    <form
      onSubmit={onSubmit} 
      data-testid="formsub">
    </form>)
  fireEvent.submit(getByTestId("formsub"))
  expect(onSubmit).toHaveBeenCalledTimes(1)
})

test('handleValidateSub return id', () => {
  const beverage = {
    sentMsg: "hello"
  }
  const handleValidateSub = vi.fn(beverage => { beverage })
  handleValidateSub()
  expect(handleValidateSub).toHaveReturned("hello")
})
