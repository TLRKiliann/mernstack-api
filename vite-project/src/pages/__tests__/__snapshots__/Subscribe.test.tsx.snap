// Vitest Snapshot v1

exports[`Subscribe toMatchSnapshot 1`] = `
<div
  className="subscribe"
  data-testid="subtest"
>
  <form
    autoComplete="off"
    className="form--subscribe"
    data-testid="formsub"
    onSubmit={[Function]}
  >
    <div
      className="div--subscribe"
    >
      <h1
        className="title--framesub"
      >
        Subscribe
      </h1>
      <label
        className="lbl--sub"
        htmlFor="sub"
      >
        Firstname
      </label>
      <input
        className="input--sub"
        name="firstName"
        onChange={[Function]}
        placeholder="firstname"
        required={true}
        type="text"
        value=""
      />
      <label
        className="lbl--sub"
        htmlFor="sub"
      >
        Lastname
      </label>
      <input
        className="input--sub"
        name="lastName"
        onChange={[Function]}
        placeholder="lastname"
        required={true}
        type="text"
        value=""
      />
      <label
        className="lbl--sub"
        htmlFor="sub"
      >
        Password
      </label>
      <input
        className="input--sub"
        name="password"
        onChange={[Function]}
        placeholder="password"
        required={true}
        type="password"
        value=""
      />
      <label
        className="lbl--sub"
        htmlFor="sub"
      >
        Age
      </label>
      <input
        className="input--sub"
        max="120"
        min="20"
        name="age"
        onChange={[Function]}
        placeholder="age"
        required={true}
        type="number"
        value={20}
      />
      <label
        className="lbl--sub"
        htmlFor="sub"
      >
        Email
      </label>
      <input
        className="input--sub"
        name="email"
        onChange={[Function]}
        placeholder="email"
        required={true}
        type="email"
        value=""
      />
      <label
        className="lbl--sub"
        htmlFor="sub"
      >
        Location
      </label>
      <input
        className="input--sub"
        name="location"
        onChange={[Function]}
        placeholder="location"
        required={true}
        type="text"
        value=""
      />
      <div
        className="div--subselect"
      >
        <label
          className="lbl--sub"
          htmlFor="gender"
        >
          Gender :
        </label>
        <select
          className="select--sub"
          id="gender"
          name="gender"
          onChange={[Function]}
          value="Male"
        >
          <option
            name="gender"
            value="Male"
          >
            Male
          </option>
          <option
            name="gender"
            value="Female"
          >
            Female
          </option>
        </select>
      </div>
    </div>
    <button
      className="btn--sub"
      type="submit"
    >
      Signup
    </button>
  </form>
</div>
`;

exports[`Subscribe toMatchSnapshot 2`] = `
<div
  className="subscribe"
  data-testid="subtest"
>
  <form
    autoComplete="off"
    className="form--subscribe"
    data-testid="formsub"
    onSubmit={[Function]}
  >
    <div
      className="div--subscribe"
    >
      <h1
        className="title--framesub"
      >
        Subscribe
      </h1>
      <label
        className="lbl--sub"
        htmlFor="sub"
      >
        Firstname
      </label>
      <input
        className="input--sub"
        name="firstName"
        onChange={[Function]}
        placeholder="firstname"
        required={true}
        type="text"
        value=""
      />
      <label
        className="lbl--sub"
        htmlFor="sub"
      >
        Lastname
      </label>
      <input
        className="input--sub"
        name="lastName"
        onChange={[Function]}
        placeholder="lastname"
        required={true}
        type="text"
        value=""
      />
      <label
        className="lbl--sub"
        htmlFor="sub"
      >
        Password
      </label>
      <input
        className="input--sub"
        name="password"
        onChange={[Function]}
        placeholder="password"
        required={true}
        type="password"
        value=""
      />
      <label
        className="lbl--sub"
        htmlFor="sub"
      >
        Age
      </label>
      <input
        className="input--sub"
        max="120"
        min="20"
        name="age"
        onChange={[Function]}
        placeholder="age"
        required={true}
        type="number"
        value={20}
      />
      <label
        className="lbl--sub"
        htmlFor="sub"
      >
        Email
      </label>
      <input
        className="input--sub"
        name="email"
        onChange={[Function]}
        placeholder="email"
        required={true}
        type="email"
        value=""
      />
      <label
        className="lbl--sub"
        htmlFor="sub"
      >
        Location
      </label>
      <input
        className="input--sub"
        name="location"
        onChange={[Function]}
        placeholder="location"
        required={true}
        type="text"
        value=""
      />
      <div
        className="div--subselect"
      >
        <label
          className="lbl--sub"
          htmlFor="gender"
        >
          Gender :
        </label>
        <select
          className="select--sub"
          id="gender"
          name="gender"
          onChange={[Function]}
          value="Male"
        >
          <option
            name="gender"
            value="Male"
          >
            Male
          </option>
          <option
            name="gender"
            value="Female"
          >
            Female
          </option>
        </select>
      </div>
    </div>
    <button
      className="btn--sub"
      type="submit"
    >
      Signup
    </button>
  </form>
</div>
`;
